File: Go-SDL2/asteroid.go
package main

import (
	"math"
	"math/rand"

	"github.com/veandco/go-sdl2/sdl"
)

type asteroid struct {
	_renderer  *sdl.Renderer
	_images    []*sdl.Texture
	_rects     []sdl.Rect
	_radii     []float64
	_xPos      float64
	_yPos      float64
	_xVel      float64
	_yVel      float64
	_angle     float64
	_angleRate float64
	_size      int
	_random    *rand.Rand
}

func newAsteroid(
	renderer *sdl.Renderer,
	random *rand.Rand,
	largeImage *sdl.Texture,
	mediumImage *sdl.Texture,
	smallImage *sdl.Texture,
	angle float64,
	xPos float64,
	yPos float64,
	xVel float64,
	yVel float64,
	size int,
) (*asteroid, error) {
	a := &asteroid{}
	var err error

	a._renderer = renderer
	a._random = random
	a._images = append(a._images, largeImage)
	a._images = append(a._images, mediumImage)
	a._images = append(a._images, smallImage)
	a._angle = angle
	a._xPos = xPos
	a._yPos = yPos
	a._xVel = xVel
	a._yVel = yVel
	a._size = size

	for size := range a._images {
		var rect sdl.Rect
		if _, _, rect.W, rect.H, err = a._images[size].Query(); err != nil {
			return a, err
		}
		a._rects = append(a._rects, rect)
		a._radii = append(a._radii, float64(rect.W/2))
	}

	if xPos == 0 || yPos == 0 {
		var centerX float64 = (windowWidth - float64(a._rects[a._size].W)) / 2
		var centerY float64 = (windowHeight - float64(a._rects[a._size].H)) / 2
		for {
			a._xPos = a._random.Float64() * windowWidth
			a._yPos = a._random.Float64() * windowHeight
			if math.Hypot(centerX-a._xPos, centerY-a._yPos) > 150 {
				break
			}
		}
	}

	a._rects[a._size].X = int32(a._xPos + 0.5)
	a._rects[a._size].Y = int32(a._yPos + 0.5)

	if angle == 0 {
		a._angle = a._random.Float64() * 360
	}

	a.reset()

	return a, err
}

func (a *asteroid) reset() {
	a._angleRate = (a._random.Float64()*2 - 1) * 60
	radian := float64(a._angle) * math.Pi / 180
	speed := a._random.Float64() * 60
	a._xVel += math.Sin(radian) * speed
	a._yVel -= math.Cos(radian) * speed
}

func (a *asteroid) split(angle float64) (*asteroid, error) {
	var err error
	xPos, yPos := a.center()
	a._size += 1
	if a._size > len(a._images)-1 {
		return nil, err
	}
	a._xPos = xPos - float64(a._rects[a._size].W)/2
	a._yPos = yPos - float64(a._rects[a._size].H)/2
	a._rects[a._size].X = int32(a._xPos + 0.5)
	a._rects[a._size].Y = int32(a._yPos + 0.5)

	angle += 90
	if angle > 360 {
		angle -= 360
	}
	newA := &asteroid{}
	if newA, err = newAsteroid(a._renderer, a._random, a._images[0], a._images[1], a._images[2], angle, a._xPos, a._yPos, a._xVel, a._yVel, a.size()); err != nil {
		return newA, err
	}

	angle += 180
	if angle > 360 {
		angle -= 360
	}
	a._angle = angle
	a.reset()

	return newA, err
}

func (a *asteroid) left() int32 {
	return int32(a._xPos)
}

func (a *asteroid) right() int32 {
	return int32(a._xPos) + a._rects[a._size].W
}

func (a *asteroid) top() int32 {
	return int32(a._yPos)
}

func (a *asteroid) bottom() int32 {
	return int32(a._yPos) + a._rects[a._size].H
}

func (a *asteroid) center() (float64, float64) {
	x := a._xPos + float64(a._rects[a._size].W)/2
	y := a._yPos + float64(a._rects[a._size].H)/2
	return x, y
}

func (a *asteroid) radius() float64 {
	return a._radii[a._size]
}

func (a *asteroid) size() int {
	return a._size
}

func (a *asteroid) update(dt float64) {
	a._angle += a._angleRate * dt
	if a._angle > 360 {
		a._angle -= 360
	} else if a._angle < 0 {
		a._angle += 360
	}

	a._xPos += a._xVel * dt
	a._yPos += a._yVel * dt

	if a.right() < 0 {
		a._xPos = windowWidth
	} else if a.left() > windowWidth {
		a._xPos = -float64(a._rects[a._size].W)
	}

	if a.bottom() < 0 {
		a._yPos = windowHeight
	} else if a.top() > windowHeight {
		a._yPos = -float64(a._rects[a._size].H)
	}

	a._rects[a._size].X = int32(a._xPos + 0.5)
	a._rects[a._size].Y = int32(a._yPos + 0.5)
}

func (a *asteroid) draw() error {
	err := a._renderer.CopyEx(a._images[a._size], nil, &a._rects[a._size], a._angle, nil, sdl.FLIP_NONE)

	return err
}

File: C-SDL2/asteroid.c
#include "asteroid.h"

bool asteroid_new(
    struct Asteroid *a,
    SDL_Renderer *renderer,
    SDL_Texture *large_image,
    SDL_Texture *medium_image,
    SDL_Texture *small_image,
    double x_pos,
    double y_pos,
    double angle,
    double x_vel,
    double y_vel,
    int size
) {
    a->renderer = renderer;
    a->images[0] = large_image;
    a->images[1] = medium_image;
    a->images[2] = small_image;
    a->x_pos = x_pos;
    a->y_pos = y_pos;
    a->angle = angle;
    a->x_vel = x_vel;
    a->y_vel = y_vel;
    a->size = size;

    for (unsigned short i = 0; i < sizeof(a->images) / sizeof(a->images[0]); i++) {
        if (SDL_QueryTexture(a->images[i], NULL, NULL, &a->rects[i].w, &a->rects[i].h)) {
            fprintf(stderr, "Error while querying texture: %s\n", SDL_GetError());
            return true;
        }
        a->radii[i] = a->rects[i].w / 2.0;
    }

    if ((int)x_pos == 0 && (int)y_pos == 0) {
        while (true) {
            double center_x = (WINDOW_WIDTH - (double)a->rects[a->size].w) / 2;
            double center_y = (WINDOW_HEIGHT - (double)a->rects[a->size].h) / 2;
            a->x_pos = rand() % WINDOW_WIDTH;
            a->y_pos = rand() % WINDOW_HEIGHT;
            if (hypot(center_x - a->x_pos, center_y - a->y_pos) > 150) {
                break;
            }
        }
    }

    if ( (int)a->angle == 0 ) {
        a->angle = rand() % 360;
    }

    if ((int)x_vel == 0 && (int)y_vel == 0) {
        asteroid_reset(a);
    } else {
        a->angle_rate = (((rand() % (200 * (a->size + 1))) / 100.0) - 1) * 60;
    }

    return false;
}

void asteroid_reset(struct Asteroid *a) {
    a->angle_rate += (((rand() % (200 * (a->size + 1))) / 100.0) - 1) * 60;
    double radian = a->angle * M_PI / 180;
    double speed = ((rand() % 100) / 100.0) * 60;
    a->x_vel += sin(radian) * speed;
    a->y_vel -= cos(radian) * speed;
}

bool asteroid_split(struct Asteroid *a, double angle) {
    double center_x = asteroid_center_x(a);
    double center_y = asteroid_center_y(a);
    a->size++;
    a->x_pos = center_x - a->radii[a->size];
    a->y_pos = center_y - a->radii[a->size];
    a->rects[a->size].x = (int)(a->x_pos + 0.5);
	a->rects[a->size].y = (int)(a->y_pos + 0.5);

    struct Asteroid *new_a = calloc(1, sizeof(struct Asteroid));
    if (new_a == NULL) {
        printf("Error in calloc of new asteroid.\n");
        return true;
    }

    if (asteroid_new(new_a,
        a->renderer,
        a->images[0],
        a->images[1],
        a->images[2],
        a->x_pos,
        a->y_pos,
        a->angle,
        a->x_vel,
        a->y_vel,
        a->size)) {
        return true;
    }

    angle += 90;
    if (angle > 360) {
        angle -= 360;
    }
    new_a->angle = angle;
    asteroid_reset(new_a);

    angle += 180;
    if (angle > 360) {
        angle -= 360;
    }
    a->angle = angle;
    asteroid_reset(a);

    if (a->next) {
        a->next->previous = new_a;
    }
    new_a->next = a->next;
    new_a->previous = a;
    a->next = new_a;

    return false;
}

double asteroid_left(struct Asteroid *a) {
    return a->x_pos;
}

double asteroid_right(struct Asteroid *a) {
    return a->x_pos + a->rects[a->size].w;
}

double asteroid_top(struct Asteroid *a) {
    return a->y_pos;
}

double asteroid_bottom(struct Asteroid *a) {
    return a->y_pos + a->rects[a->size].h;
}

double asteroid_center_x(struct Asteroid *a) {
    return a->x_pos + a->rects[a->size].w / 2.0;
}

double asteroid_center_y(struct Asteroid *a) {
    return a->y_pos + a->rects[a->size].h / 2.0;
}

double asteriod_radius(struct Asteroid *a) {
    return a->radii[a->size];
}

void asteroid_update(struct Asteroid *a, double dt) {
    a->angle += a->angle_rate * dt;
    if (a->angle > 360) {
        a->angle -= 360;
    } else if (a->angle < 0) {
        a->angle += 360;
    }

    a->x_pos += a->x_vel * dt;
    a->y_pos += a->y_vel * dt;

    if (a->x_pos + a->rects[a->size].w < 0) {
        a->x_pos = WINDOW_WIDTH;
    } else if (a->x_pos > WINDOW_WIDTH) {
        a->x_pos = -a->rects[a->size].w;
    }

    if (a->y_pos + a->rects[a->size].h < 0) {
        a->y_pos = WINDOW_HEIGHT;
    } else if (a->y_pos > WINDOW_HEIGHT) {
        a->y_pos = -a->rects[a->size].h;
    }

    a->rects[a->size].x = (int)(a->x_pos + 0.5);
	a->rects[a->size].y = (int)(a->y_pos + 0.5);
}

bool asteroid_draw(struct Asteroid *a) {
    if ( SDL_RenderCopyEx(a->renderer, a->images[a->size], NULL, &a->rects[a->size], a->angle, NULL, SDL_FLIP_NONE)) {
        fprintf(stderr, "Error while rendering texture: %s\n", SDL_GetError());
        return true;
    }

    return false;
}